### PR TITLE
implement memory utilization fixture

### PR DESCRIPTION
**Pull request by lipxu**
### Description of PR

Summary:
Fixes # (issue)
27216861

### Type of change

- [ ] Bug fix
- [ ] Testbed and Framework(new/improvement)
- [ ] Test case(new/improvement)

### Back port request
- [ ] 202012
- [ ] 202205
- [ ] 202305
- [ ] 202311
- [ ] 202405

#### What is the motivation for this PR?
Test gap for memory leak

#### How did you do it?
Introduced a new fixture to collect the memory information before and after the test case.
Then compare the memory information to confirm that it has not exceeded the high memory usage threshold and that no memory leaks have occurred.

#### How did you verify/test it?
Run case locally.
Set a fake threshold, check the result.
Run nightly pipeline.

### CODE REVIEW

File: tests/common/plugins/memory_utilization/memory_utilization_common.json [hunk 1]
```diff
@@ -0,0 +1,15 @@
+{
+  "COMMON": [
+    {
+      "name": "monit",
+      "cmd": "sudo monit status",
+      "memory_params": {
+        "memory_usage": {
+          "memory_increase_threshold": 5,
+          "memory_high_threshold": 70
```
Comment (wsycqyz) [inline tests/common/plugins/memory_utilization/memory_utilization_common.json:9]: Just question. Any reason why it is 70 not 80?

Comment (lipxu) [reply tests/common/plugins/memory_utilization/memory_utilization_common.json:9]: Thanks for asking, Currently, the SONiC Memory Utilization Threshold is 70% in ICM. So we set 70 here in public branch.

File: tests/common/plugins/memory_utilization/README.md [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+    {
+      "name": "monit",
+      "cmd": "sudo monit status",
+      "memory_params": {
+        "memory_usage": {
+          "memory_increase_threshold": 5,
+          "memory_high_threshold": 70
+        }
+      },
+      "memory_check": "parse_monit_status_output"
```
[the remark this answers is not in the document]
Comment (lipxu) [reply tests/common/plugins/memory_utilization/README.md:71]: Thanks for the reminder, updated the README adding more descriptions, thanks

File: tests/common/plugins/memory_utilization/README.md [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+#### To skip memory_utilization for:
+- all test cases - use pytest command line option ```--memory_utilization```
+- specific test case: mark test case with ```@pytest.mark.memory_utilization``` decorator. Example is shown below.
+    ```python
+    pytestmark = [
+        pytest.mark.disable_memory_utilization
+    ]
+    ```
+
+#### Example of memory items configuration in json file
```
Comment (StormLiangMS) [inline tests/common/plugins/memory_utilization/README.md:55]: hi @lipxu  could we list the command we recommend for the memory usage / leak check? And how they compare when it is exceeded the threshold with examples?

File: tests/common/plugins/memory_utilization/README.md [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+
+The purpose of the current feature is to verify that memory resources do not increase during test runs and do not exceed the high memory usage threshold.
+
+### Module Design
+Newly introduced a plugin "memory_utilization" and config files.
+
+#### Config files
+Config files, including common file and platform dependence file.
+- **memory_utilization_common.json**: Common configurations for memory utilization.
+- **memory_utilization_dependence.json**: Dependency configruations for memory utilization, currently supporting special configurations for specific HwSku devices. This configuration file should be left empty in the public branch, with special configurations added in internal branch.
```
Comment (StormLiangMS) [inline tests/common/plugins/memory_utilization/README.md:15]: @lipxu  could we give detailed example about the hwsku configuration?
could we have customized check per testcase?

**Comment by StormLiangMS** [comment]
@lipxu could we have an example to give when this check passed / failed?

File: tests/common/plugins/memory_utilization/README.md (hunk 3 above)
Comment (lipxu) [reply tests/common/plugins/memory_utilization/README.md:65]: Sure, updated the README with additional examples for current support memory items. thanks

File: tests/common/plugins/memory_utilization/README.md (hunk 4 above)
Comment (lipxu) [reply tests/common/plugins/memory_utilization/README.md]: Sure, had updated the README, and add the example for hwsku configuration

**Comment by lipxu** [comment]
> @lipxu could we have an example to give when this check passed / failed?

Sure, added the examples for supported memory items.

File: tests/common/plugins/memory_utilization/README.md [hunk 5]
```diff
@@ -0,0 +1,573 @@
...
+Newly introduced a plugin "memory_utilization" and config files.
+
+#### Config files
+Config files, including both common file and platform dependence file.
+- **memory_utilization_common.json**: Common configurations for memory utilization.
+  - The file is in JSON format and is defined in the public branch.
+  - It includes "COMMON" key which holds all publicly defined common memory items.
+- **memory_utilization_dependence.json**: Dependency configruations for memory utilization, currently supporting special configurations for specific HwSku devices.
+  - The dependency file is also in JSON format and is used in internal branch.
+  - It alos includes a 'COMMON' key which holds all internal common memory items. If an internal memory item has the same name as an item in the public common configuration file, it will overwrite the public common item.
```
Comment (ZhaohuiS) [inline tests/common/plugins/memory_utilization/README.md:19]: @lipxu typo, also

File: tests/common/plugins/memory_utilization/README.md [hunk 6]
```diff
@@ -0,0 +1,573 @@
...
+In the root conftest there is an implemented "memory_utilization" pytest fixture that starts automatically for all test cases.
+The main flow of the fixture is as follows:
+- memory_utilization collects memory information before the test case starts.
+- memory_utilization collects memory information after the test case finishes.
+- memory_utilization compares DUT memory usage and displays the results.
+- if memory_utilization finds any exceeded thresholds for high memory usage or memory increase, it will display the result and pytest will generate an 'error'.
+
+#### To skip memory_utilization for:
+- all test cases - use pytest command line option ```--memory_utilization```
+- specific test case: mark test case with ```@pytest.mark.memory_utilization``` decorator. Example is shown below.
```
Comment (ZhaohuiS) [inline tests/common/plugins/memory_utilization/README.md:58]: @lipxu should be disable_memory_utilization decorator?

File: tests/common/plugins/memory_utilization/README.md [hunk 7]
```diff
@@ -0,0 +1,573 @@
...
+##### memory_utilization fixture
+In the root conftest there is an implemented "memory_utilization" pytest fixture that starts automatically for all test cases.
+The main flow of the fixture is as follows:
+- memory_utilization collects memory information before the test case starts.
+- memory_utilization collects memory information after the test case finishes.
+- memory_utilization compares DUT memory usage and displays the results.
+- if memory_utilization finds any exceeded thresholds for high memory usage or memory increase, it will display the result and pytest will generate an 'error'.
+
+#### To skip memory_utilization for:
+- all test cases - use pytest command line option ```--memory_utilization```
```
Comment (ZhaohuiS) [inline tests/common/plugins/memory_utilization/README.md:57]: @lipxu a little confusion here, is --memory_utilization  disable or enable the memory check?

File: tests/common/plugins/memory_utilization/README.md (hunk 5 above)
Comment (lipxu) [reply tests/common/plugins/memory_utilization/README.md:19]: thanks for reminder, fixed.

File: tests/common/plugins/memory_utilization/README.md (hunk 6 above)
Comment (lipxu) [reply tests/common/plugins/memory_utilization/README.md:58]: thanks for reminder, fixed.

File: tests/common/plugins/memory_utilization/README.md (hunk 7 above)
Comment (lipxu) [reply tests/common/plugins/memory_utilization/README.md:57]: Thanks for the reminder, fixed. 
memory_utilization , Default is enabled, if we want to disable the check, we can use disable_memory_utilization to skip the memory check.

File: tests/common/plugins/memory_utilization/README.md [hunk 8]
```diff
@@ -0,0 +1,573 @@
+### Scope
```
Comment (StormLiangMS) [inline tests/common/plugins/memory_utilization/README.md:1]: @lipxu  could we have a section as use case, and give introduction how to use this fixture?

1. To define a memory check globally. 
2. To define a memory check per HWSKU
3. To define a memory check per test module/test case
4. When different memory check defined for global, HWSKU, test cases, how them interact?

Comment (lipxu) [reply tests/common/plugins/memory_utilization/README.md:1]: Sure, updated the README , add "Example Use Cases for memory items" section, thanks for reminder, please review again.

**Comment by StormLiangMS** [comment]
hi @lipxu should we consider when this fixture can kick in? if it is after teardown, it may not work as expected since all process could be already reloaded without memory leak evidence.

**Comment by lipxu** [comment]
> hi @lipxu should we consider when this fixture can kick in? if it is after teardown, it may not work as expected since all process could be already reloaded without memory leak evidence.

I've updated the branch by adding the following two functions and adjusted the setup to use "trylast" and teardown to use "tryfirst". This will allow memory utilization to collect the memory information before and after the test case.  ensuring accurate capture memory usage of the test case.

@pytest.hookimpl(trylast=True)
def pytest_runtest_setup(item):

@pytest.hookimpl(tryfirst=True)
def pytest_runtest_teardown(item, nextitem):

**Review by StormLiangMS — APPROVED** [review]
LGTM

**Comment by StormLiangMS** [comment]
hi @lipxu LGTM, could you work with other reviewers to close the comment before we get this merged?

**Review by ZhaohuiS — APPROVED** [review]

**Comment by dhanasekar-arista** [comment]
> capture memory usage of the test case

@StormLiangMS  @lipxu  I have a concern on this block 
```
@pytest.hookimpl(trylast=True)
def pytest_runtest_setup(item):

@pytest.hookimpl(tryfirst=True)
def pytest_runtest_teardown(item, nextitem):
```

There may be fixtures managing configuration setup and cleanup around the test. In such cases, the memory checker should run only after all post-test cleanup has been completed. Else memory checker wouldn't give us expected results.

**Pull request merged**